File: webide/src/main/java/fiveguys/webide/api/user/service/UserService.java
package fiveguys.webide.api.user.service;

import fiveguys.webide.api.user.dto.request.LoginRequest;
import fiveguys.webide.api.user.dto.request.SignupRequest;
import fiveguys.webide.api.user.dto.response.LoginResponse;
import fiveguys.webide.common.error.ErrorCode;
import fiveguys.webide.common.error.GlobalException;
import fiveguys.webide.domain.user.User;
import fiveguys.webide.domain.user.UserRole;
import fiveguys.webide.repository.user.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import static fiveguys.webide.common.error.ErrorCode.RETRY_EMAIL_AUTH;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class UserService {
    private final UserRepository userRepository;
    private final BCryptPasswordEncoder bCryptPasswordEncoder;
    private final AuthenticationManager authenticationManager;

    @Transactional
    public void signup(SignupRequest request) {
        if (!request.isValid()) {
            throw new GlobalException(RETRY_EMAIL_AUTH);
        }
        User user = request.toEntity(bCryptPasswordEncoder.encode(request.getPassword()));
        userRepository.save(user);
    }

    @Transactional
    public LoginResponse login(LoginRequest request) {
        UsernamePasswordAuthenticationToken authenticationToken = new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword());
        Authentication authentication = authenticationManager.authenticate(authenticationToken);
        SecurityContextHolder.getContext().setAuthentication(authentication);
        return null;
    }
}

File: webide/src/main/kotlin/fiveguys/webide/user/service/UserService.kt
package fiveguys.webide.user.service

import fiveguys.webide.common.aop.Trace
import fiveguys.webide.config.jwt.JwtUtils
import fiveguys.webide.user.domain.UserRepository
import fiveguys.webide.user.dto.request.LoginRequest
import fiveguys.webide.user.dto.request.SignupRequest
import fiveguys.webide.user.dto.response.LoginUserResponse
import fiveguys.webide.common.error.ErrorCode
import fiveguys.webide.common.error.GlobalException
import fiveguys.webide.config.auth.LoginUser
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class UserService(
    private val userRepository: UserRepository,
    private val bCryptPasswordEncoder: BCryptPasswordEncoder,
    private val authenticationManager: AuthenticationManager,
    private val jwtUtils: JwtUtils,
) {

    @Transactional
    @Trace
    fun signup(request: SignupRequest) {
        if (userRepository.existsByEmail(request.email)) {
            throw GlobalException(ErrorCode.EXIST_EMAIL)
        }
        if (userRepository.existsByNickname(request.nickname)) {
            throw GlobalException(ErrorCode.EXIST_NICKNAME)
        }
        val user = request.toEntity(bCryptPasswordEncoder.encode(request.password))
        userRepository.save(user)
    }

    @Transactional
    fun login(request: LoginRequest): LoginUserResponse {
        val authenticationToken = UsernamePasswordAuthenticationToken(request.email, request.password)
        val authentication = authenticationManager.authenticate(authenticationToken)
        SecurityContextHolder.getContext().authentication = authentication
        val loginUser = authentication.principal as LoginUser
        val accessToken = jwtUtils.generateAccessTokenFromLoginUser(loginUser)
        return LoginUserResponse(
            loginUser,
            accessToken,
            "refreshToken"
        )
    }
}
